linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,247 @@
+use ndarray_stats::DeviationExt;
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+use std::cmp::{max, Ordering};
+use std::collections::HashSet;
+
+use linfa::traits::Transformer;
+use linfa::Float;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+/// OPTICS (Ordering Points To Identify Clustering Structure) is a clustering algorithm that
+/// doesn't explicitly cluster the data but instead creates an "augmented ordering" of the dataset
+/// representing it's density-based clustering structure. This ordering contains information which
+/// is equivalent to the density-based clusterings and can then be used for automatic and
+/// interactive cluster analysis.
+///
+/// OPTICS cluster analysis can be used to derive clusters equivalent to the output of other
+/// clustering algorithms such as DBSCAN. However, due to it's more complicated neighborhood
+/// queries it typically has a higher computational cost that other more specific algorithms.
+pub struct Optics;
+
+#[derive(Clone, Debug)]
+// This is a struct as in future may want to implement methods on it to get certain metrics from
+// the optics cluster distances.
+pub struct OpticsResult<'a, F: Float> {
+    /// A list of the samples in the dataset sorted and with their reachability and core distances
+    /// computed.
+    pub orderings: Vec<Sample<'a, F>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Sample<'a, F: Float> {
+    /// Index of the sample in the dataset.
+    pub index: usize,
+    /// A reference to the datum.
+    pub observation: ArrayView<'a, F, Ix1>,
+    /// The reachability distance of a sample is the distance between the point and it's cluster
+    /// core or another point whichever is larger.
+    pub reachability_distance: Option<f64>,
+    /// The distance to the nth closest point where n is the minimum points to form a cluster.
+    pub core_distance: Option<f64>,
+}
+
+#[derive(Clone)]
+struct Neighbor<'a, F: Float> {
+    /// Index of the observation in the dataset
+    index: usize,
+    /// The observation
+    observation: ArrayView<'a, F, Ix1>,
+    /// Whether this point has been processed
+    processed: bool,
+    /// The core distance
+    c_distance: Option<FloatOrd<f64>>,
+    /// The reachability distance
+    r_distance: Option<FloatOrd<f64>>,
+}
+
+impl<'a, F: Float> Neighbor<'a, F> {
+    /// Create a new neighbor
+    fn new(index: usize, observation: ArrayView<'a, F, Ix1>) -> Self {
+        Self {
+            index,
+            observation,
+            processed: false,
+            c_distance: None,
+            r_distance: None,
+        }
+    }
+
+    /// Set the core distance given the minimum points in a cluster and the points neighbors
+    fn set_core_distance(&mut self, min_pts: usize, neighbors: &[Neighbor<'a, F>]) {
+        self.c_distance = neighbors
```

Comment:
But if the point isn't a core point the distance should be `None`, I figured this was a reasonable way to avoid misuse of the function without sacrificing performance

Reply:
ah I see, perhaps you can use the `slice::get` syntax:
```rust
self.c_distance = neighbors.get(min_pts-1)
    .and_then(|x| self.observation.l2_dist(x))
    .ok();
```
A possible error originating from the `l2_dist` is converted into `None` with the last line